worker/secretsdrainworker/worker.go

```diff
@@ -108,30 +114,74 @@ func (w *Worker) loop() (err error) {
 				return errors.New("secret backend changed watch closed")
 			}
 			w.config.Logger.Debugf("got new secret backend")
-
-			secrets, err := w.config.SecretsDrainFacade.GetSecretsToDrain()
-			if err != nil {
-				return errors.Trace(err)
-			}
-			if len(secrets) == 0 {
-				w.config.Logger.Debugf("no secrets to drain")
-				continue
-			}
-			w.config.Logger.Debugf("got %d secrets to drain", len(secrets))
-			backends, err := w.config.SecretsBackendGetter()
-			if err != nil {
-				return errors.Trace(err)
-			}
-			for _, md := range secrets {
-				if err := w.drainSecret(md, backends); err != nil {
+			for {
+				switch err := w.drainSecrets(); {
+				case err == nil:
+					continue waitforchanges
+				case errors.Is(err, leadership.ErrLeadershipChanged):
+					// If leadership changes during the drain operation,
+					// we need to finish up and start again.
+					w.config.Logger.Warningf("leadership changed, restarting drain operation")
+					continue
+				default:
 					return errors.Trace(err)
 				}
 			}
 		}
 	}
 }
 
-func (w *Worker) drainSecret(md coresecrets.SecretMetadataForDrain, client jujusecrets.BackendsClient) (err error) {
+// scopedContext returns a context that is in the scope of the worker lifetime.
+// It returns a cancellable context that is cancelled when the action has
+// completed.
+func (w *Worker) scopedContext() (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+	return w.catacomb.Context(ctx), cancel
+}
+
+// drainSecrets queries the secrets owned by the unit and if the unit is
+// the leader, this will include any application owned secrets.
+// If leadership changes during the draining operation, an error satisfying
+// [leadershipworker.ErrLeadershipChanged] is returned.
+func (w *Worker) drainSecrets() error {
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	leadershipTracker := w.config.LeadershipTrackerFunc()
+	drainErr := leadershipTracker.WithoutLeadershipChange(ctx, func(ctx context.Context) error {
+		secrets, err := w.config.SecretsDrainFacade.GetSecretsToDrain()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(secrets) == 0 {
+			w.config.Logger.Debugf("no secrets to drain")
+			return nil
+		}
+		w.config.Logger.Debugf("got %d secrets to drain", len(secrets))
+		backends, err := w.config.SecretsBackendGetter()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, md := range secrets {
+			if err := w.drainSecret(ctx, md, backends); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	})
```

Comment:
if drainErr is context cancelled error, we should check if the catacomb/tomb is dying then return ErrDying() instead. See dbaccessor worker or the undertaker worker's handling of `context.Canceled`.

Reply:
yup